server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1198,63 +1197,56 @@ private static Collection<String> innerResolve(Context context, List<String> exp
+                if (Regex.isSimpleMatchPattern(expression)) {
+                    wildcardSeen = true;
+                    Stream<IndexAbstraction> matchingResources = matchResourcesToWildcard(context, expression);
+                    Stream<String> matchingOpenClosedNames = expandToOpenClosed(context, matchingResources);
+                    AtomicBoolean emptyWildcardExpansion = new AtomicBoolean(false);
+                    if (context.getOptions().allowNoIndices() == false) {
+                        emptyWildcardExpansion.set(true);
+                        matchingOpenClosedNames = matchingOpenClosedNames.peek(x -> emptyWildcardExpansion.set(false));
+                    }
                     if (result == null) {
                         // add all the previous expressions because they exist but were not added, as an optimisation
                         result = new HashSet<>(expressions.subList(0, i));
                     }
-                    /* An expression does not exist when:
-                       * it should be treated as exclusion because it starts with "-"
-                       * it should be treated as a wildcard (inclusion or exclusion) because it contains a "*"
-                       * it genuinely does not exist or is a datastream or an alias and the request type and options disallow it
-                     */
-                    final boolean add;
-                    if (expression.charAt(0) == '-' && wildcardSeen) {
-                        add = false;
-                        expression = expression.substring(1);
+                    if (isExclusion) {
+                        matchingOpenClosedNames.forEachOrdered(result::remove);
                     } else {
-                        add = true;
+                        matchingOpenClosedNames.forEachOrdered(result::add);
                     }
-                    if (Regex.isSimpleMatchPattern(expression) == false) {
-                        if (add) {
-                            // missing expression that is neither an exclusion nor a wildcard
-                            // this must be a name for a resource that genuinely does not exist
-                            // TODO investigate if this check can be moved outside the wildcard resolver
-                            if (context.getOptions().ignoreUnavailable() == false) {
-                                throw missingExpressionException.get();
-                            }
-                            result.add(expression);
-                        } else {
-                            result.remove(expression);
+                    if (emptyWildcardExpansion.get()) {
+                        throw indexNotFoundException(expression);
+                    }
+                } else {
+                    if (isExclusion) {
+                        if (result == null) {
+                            // add all the previous expressions because they exist but were not added, as an optimisation
```

Comment:
I don't understand this optimization. Can't we just create the Collection outside the loop and use it normally ?

Reply:
I can explain the optimization, but that doesn't mean that I'm endorsing it (it's not my creation).

If the expression consists exclusively of explicit names that all exist, then it avoids allocating another collection, and the wildcard resolver here is effectively a noop. This might be a somewhat frequent path, when the wildcard resolver in the Security filter expanded to explicit names that are all valid and exist.

Reply:
I understand this is not your code nor relevant to this PR...so this is likely a follow up PR if you agree. 

I dislike that optimization for the following reasons:
* It is a single allocation. It might be a frequent path, but it's not hot in the sense of a tight loop. 
* It is a bit of code smell to flip between a List and Set based on the criteria here. It may be a non-issue but it is odd that duplicated expressions are de-dupped when a wildcard is present. It hints at subtle behavior constraints that don't exist.
* It is complicates the code and even with the comments kinda begs the question to why (see above about subtle behavior constraints) 
* micro optimizations like this can often have the opposite effect. For example, how expensive is the .subList and what is the primary workflows (now and after combining with security)?  Primary workflow is hard to answer, and more expensive requires a proper micro benchmark.

Just food for thought, thanks for the explanation.

Reply:
I agree with the downsides. It is probably not an optimization that we should concern ourselves with (I think there are more egregious expression copying elsewhere). I'm oscilating between removing it altogether, or having a separate pass over the expression to implement it there and not have it mixed with the wildcard logic here. Leaning to remove in a follow-up....

Thanks for the advice and the thought!